### DISCUSSION TITLE
Help debugging assertion errors fn OnApplyAttackPlayer

### DISCUSSION THREAD

**Discussion by danlangford** [pappde/bmai#11] [Q&A]
so i have 3 games that are stuck right now. They are saying "assertion failed" and give a line number from the code. Im going to try to dig into these errors and thought I would report on my progress here in case you were able to give some hints or helps. 

On the first one the game hasn't quite started, we are in the preround selecting swing dice sizes. 
```
Assertion failed: (g_attack_type[_move.m_attack]==BME_ATTACK_TYPE_1_1 || g_attack_type[_move.m_attack]==BME_ATTACK_TYPE_N_1), function OnApplyAttackPlayer, file src/bmai.cpp, line 872.
```

The Second problem comes from 2 games. Has to do with Reserve dice and I will get those inputs and posted here in a bit. 
```
Assertion failed: (i>0 || m_sides[i]>0), function Roll, file src/bmai.cpp, line 804.
```

edit: so i am learning i dont on this machine have good tools set up for debugging and stepping through code and seeing stack traces. still new to the C++ dev cycle. i come from java.

**Comment by danlangford** [comment]
On the first one the game hasn't quite started, we are in the preround selecting swing dice sizes. 
INPUT
```
preround
player 0 3 0
W
om34
T
player 1 5 0
^10
s11
d12
z13
X
ply 3
max_sims 100
min_sims 5
maxbranch 400
getaction

```

OUTPUT

```
% cat ~/Downloads/69369.txt | ./bmai
BMAI: the Button Men AI
Copyright (c) 2001-2020, Denis Papp.
For information, contact Denis Papp, denis@accessdenied.net
target wins set to 3
s 0.0 Dice (0)0:0 (480000)34:0 (0)0:0
s 0.0 Dice (1)10:0 (20)11:0 (80)12:0 (10)13:0 (0)0:0
Setting max ply to 3
Setting max # simulations to 100
Setting min # simulations to 5
Setting max branch to 400
l1 p0 Valid SetSwing 99 Max 80
l1 p0 Valid SetSwing 80 Sims 5
l1 p0 - swing/option T 2  W 4
l2 p1 Valid SetSwing 17 Sims 11
l2 p1 - swing/option X 4
Assertion failed: (g_attack_type[_move.m_attack]==BME_ATTACK_TYPE_1_1 || g_attack_type[_move.m_attack]==BME_ATTACK_TYPE_N_1), function OnApplyAttackPlayer, file src/bmai.cpp, line 872.
zsh: done       cat ~/Downloads/69369.txt |
zsh: abort      ./bmai
```


-------



code in question is here:

```
	// MORPHING - assume same size as target
	if (HasProperty(BME_PROPERTY_MORPHING))
	{
		BM_ASSERT(g_attack_type[_move.m_attack]==BME_ATTACK_TYPE_1_1 || g_attack_type[_move.m_attack]==BME_ATTACK_TYPE_N_1);
		BMC_Player *target = _move.m_game->GetPlayer(_move.m_target_player);

		BM_ASSERT(Dice()==1);

		_owner->OnDieSidesChanging(this);
		m_sides_max = m_sides[0] = target->GetDie(_move.m_target)->GetSidesMax();
		_owner->OnDieSidesChanged(this);
	}
```

the morphing die is `om34`

**Reply by danlangford** [reply]
```
m_attack=6 g_attack_type[_move.m_attack]=3 BME_ATTACK_TYPE_1_1=0 BME_ATTACK_TYPE_N_1=1
```

so here is what i think is happening
```
enum BME_ATTACK
{
	BME_ATTACK_FIRST = 0,			//0. 
	BME_ATTACK_POWER = BME_ATTACK_FIRST,	//0.  1 -> 1
	BME_ATTACK_SKILL,			//1.  N -> 1
	BME_ATTACK_BERSERK,			//2.  N -> 1
	BME_ATTACK_SPEED,			//3.  1 -> N
	BME_ATTACK_TRIP,			//4.  1 -> 1
	BME_ATTACK_SHADOW,			//5.  1 -> 1
	BME_ATTACK_INVALID,			//6.  0 -> 0 (like pass)
	BME_ATTACK_MAX				//7.
};
```
i dont know exactly what caused the ATTACK INVALID. the 6. which is a 0->0 attack. and when that is mapped to attack type …. 

```
BME_ATTACK_TYPE	g_attack_type[BME_ATTACK_MAX] =
{
	BME_ATTACK_TYPE_1_1,		//0.
	BME_ATTACK_TYPE_N_1,		//1.
	BME_ATTACK_TYPE_1_N,		//2.
	BME_ATTACK_TYPE_1_N,		//3.
	BME_ATTACK_TYPE_1_1,		//4.
	BME_ATTACK_TYPE_1_1,		//5.
	BME_ATTACK_TYPE_0_0,		//6.
};
```

we actually produce `BME_ATTACK_TYPE_0_0` which is not desirable. that's what fails the assert 


edit: for posterity ill also place in the `enum BME_ATTACK_TYPE`

```
enum BME_ATTACK_TYPE
{
	BME_ATTACK_TYPE_1_1,		//0.
	BME_ATTACK_TYPE_N_1,		//1.
	BME_ATTACK_TYPE_1_N,		//2.
	BME_ATTACK_TYPE_0_0,		//3.
	BME_ATTACK_TYPE_MAX		//4.
};
```

**Reply by danlangford** [reply]
i learned that the INVALID attack type comes from a PASS game action. 

so i am learning i dont on this machine have good tools set up for debugging and stepping through code and seeing stack traces. still new to the C++ dev cycle. i come from java. but I did learn that the possible call stacks come through 2 places

1. a switch that guarantees we will have `BME_ATTACK_TYPE_1_1` or `BME_ATTACK_TYPE_1_N` or another case `BME_ATTACK_TYPE_N_1` and then I guess because its a morphing the only possible attack that could have spawned this code would be the two types in the assert

2. also the following code can (and is) calling the method. the di in question is not only morphing it is also ORNERY. 

```
         // ORNERY: all ornery dice on attacker must reroll (whether or not attacked)
	for (i=0; i<attacker->GetAvailableDice(); i++)
	{
		att_die = attacker->GetDie(i);
		if (att_die->HasProperty(BME_PROPERTY_ORNERY))
			att_die->OnApplyAttackPlayer(_move,attacker,false);	// false means not _actually_attacking
	}
``` 

the code doesn't make sure it only makes calls with a subset of BME_ATTACK_TYPEs. which is appropriate because ORNERY are almost always rolling. 

so this brings me to two questions. 

1. should the ORNERY block there be allowed to fire if the attack type is invalid?
2. should the MORPHING block of code be reacting differently if `_actually_attacking==false`

these I will consider and research

**Reply by danlangford** [reply]
> **Ornery (o)** Ornery dice reroll every time the player makes any attack - whether the Ornery dice participated in it or not. The only time they don't reroll is if the player passes, making no attack whatsoever.

edit: see issue #14

**Reply by pappde** [reply]
DEBUGGING
I started a separate discussion about this and project configuration.

OTHER ISSUES
- Twin Dice: I fixed the Dice() method incorrectly returning "bool". It should return integer type. This may have worked in older compilers where it treated a bool as an unsigned char, but won't work in compilers that treat it as "true/false".
- Compiler warnings: fixed a few loss of precision from implicit cast warnings.

**Comment by danlangford** [comment]
this is a placeholder for coming and discussing in a moment my other problem during Reserve dice selection

```
Assertion failed: (i>0 || m_sides[i]>0), function Roll, file src/bmai.cpp, line 804.
```

edit: ok here we go

INPUT
```
game 3
reserve
player 0 8 0
qT
qW
qX
qZ
rnR
rzS
rpU
rfV
player 1 5 0
B8
H8
p8
z8
zU?-30
ply 3
max_sims 100
min_sims 5
maxbranch 400
getaction
quit
```

OUTPUT
```
% cat ~/Downloads/71297.txt | ./bmai
BMAI: the Button Men AI
Copyright (c) 2001-2020, Denis Papp.
For information, contact Denis Papp, denis@accessdenied.net
target wins set to 3
s 0.0 Dice (4)0:0 (4)0:0 (4)0:0 (4)0:0
s 0.0 Dice (40)8:0 (10000)8:0 (100)8:0 (10)8:0 (410)30:0
Setting max ply to 3
Setting max # simulations to 100
Setting min # simulations to 5
Setting max branch to 400
l2 p0 Valid SetSwing 45441 Max 80
l2 p0 Valid SetSwing 80 Sims 3
l2 p0 - swing/option T 2  W 4  X 4  Z 4
Assertion failed: (i>0 || m_sides[i]>0), function Roll, file src/bmai.cpp, line 804.
```

**Reply by jl8e** [reply]
Are there reserve swing dice in these games? BMAI had problems with them.

**Reply by pappde** [reply]
ASSERT #2 
Do you have a test input to trigger that assert? 

As to your question, I imagine any type of dice could be "twin", "reserve" and "swing" - at a glance it seems the parsing could probably a little more robust. Regardless, assuming it parses correctly, I wouldn't be surprised if some combinations are untested and not properly handled in the code.

Somewhat related, in the previously noted commit I fixed a potential problem with the Dice() method around that for loop. It should return either "1" or "2", but was returning as "bool" which might have been interpreted incorrectly as "0" or "1".  That said, it seems unlikely to have led to this particular assert tripping, but might want to fix and run again.

In BMC_DieData
`INT			Dice()				 { return (m_properties & BME_PROPERTY_TWIN) ? 2 : 1; }`

**Reply by danlangford** [reply]
oh hey @jl8e i didnt know you were lurking over here :-) 

> Are there reserve swing dice in these games? BMAI had problems with them.

there are reserve Swing dice

edit: see #16

**Comment by pappde** [comment]
ASSERT # 1
I see this is an "ornery morphing" die.  So either it is an invalid assumption that the two properties wouldn't be combined, or it is not a supported combination, or it is simply a bug particular to one.

Looking closer at the code, it looks to be the latter. The "MORPHING" code is stating that it only supports 1:1 and N:1 attack types, but we are looking at a "pass" (0:0). In that case, we shouldn't be entering the block at all. I will submit a fix.

**Reply by pappde** [reply]
See commit d8913b0c6276c0812271ff22b0562f6ebd6b03b4 [d8913b0] in main.  Not sure if there's an easy way to tag a commit from a discussion, or tag a commit TO a discussion in the commit description

The specific fix bmai.cpp line ~866:
`if (HasProperty(BME_PROPERTY_MORPHING) && _move.m_attack != BME_ATTACK_INVALID)`

However, the commit includes a couple of other bug fixes

**Reply by danlangford** [reply]
I was going to reply but got caught up in other things. I actually think the bug to fix is "ornery die should not reroll if no attack was made" so the fix I made but didn't commit or comment yet was out to not for-loop spin through dice rerolling the ornery dice if no attack was made (pass, 0:0) 
edit: see issue #14

**Reply by danlangford** [reply]
@hamstercrack this is your project you are welcome to commit to main. HOWEVER if you want to create a branch and a Pull Request github actions will auto build the PR on all three Operating Systems to help validate the change before its merged into main. we can even work towards a good automated testing scenario in the future to validate even more logic before changes are merged into main

**Reply by pappde** [reply]
That sounds awesome...  EDIT: I just saw the ".github" folder so I see how it is configured.
